provider/linode/linode_discover.go

```diff
@@ -0,0 +1,163 @@
+    region:       The Linode region to filter on
+    tag_name:     The tag name to filter on
+    address_type: "private_v4", "public_v4", "private_v6" or "public_v6". (default: "private_v4")
+
+    Variables can also be provided by environment variables:
+    export LINODE_TOKEN for api_token
+`
+}
+
+func (p *Provider) Addrs(args map[string]string, l *log.Logger) ([]string, error) {
+	if args["provider"] != "linode" {
+		return nil, fmt.Errorf("discover-linode: invalid provider " + args["provider"])
+	}
+
+	if l == nil {
+		l = log.New(ioutil.Discard, "", 0)
+	}
+
+	addressType := args["address_type"]
+	region := args["region"]
+	tagName := args["tag_name"]
+	apiToken := argsOrEnv(args, "api_token", "LINODE_TOKEN")
+	l.Printf("[DEBUG] discover-linode: Using address_type=%s region=%s tag_name=%s", addressType, region, tagName)
+
+	client := getLinodeClient(p.userAgent, apiToken)
+
+	filters := Filter{
+		Region: "",
+		Tag:    "",
+	}
+
+	if region != "" {
+		filters.Region = region
+	}
+	if tagName != "" {
+		filters.Tag = tagName
+	}
+
+	jsonFilters, _ := json.Marshal(filters)
+	filterOpt := linodego.ListOptions{Filter: string(jsonFilters)}
+
+	linodes, err := client.ListInstances(context.Background(), &filterOpt)
+	if err != nil {
+		return nil, fmt.Errorf("discover-linode: Fetching Linode instances failed: %s", err)
+	}
+
+	var addrs []string
+	for _, linode := range linodes {
+		if linodeAddrs, err := getLinodeAddresses(client, addressType, linode); err == nil {
+			addrs = append(addrs, linodeAddrs...)
+		} else {
+			return nil, err
+		}
+	}
+
+	return addrs, nil
+}
+
+func getLinodeAddresses(client linodego.Client, addressType string, linode linodego.Instance) (addresses []string, err error) {
+	switch addressType {
+	case "public_v4":
+		for _, ip := range linode.IPv4 {
+			if !privateIP(*ip) {
+				addresses = append(addresses, ip.String())
+			}
+		}
+	case "public_v6":
+		v6Addr := strings.SplitN(linode.IPv6, "/", 2)
+		if len(v6Addr) > 0 {
+			addresses = append(addresses, v6Addr[0])
+		}
+	case "private_v6":
+		var addr *linodego.InstanceIPAddressResponse
+		if addr, err = client.GetInstanceIPAddresses(context.Background(), linode.ID); err == nil {
+			if addr.IPv6.LinkLocal.Address != "" {
+				addresses = append(addresses, addr.IPv6.LinkLocal.Address)
+			}
+		} else {
+			err = fmt.Errorf("discover-linode: Fetching Linode IP address for instance %v failed: %s", linode.ID, err)
+		}
+	// private_v4
+	default:
+		for _, ip := range linode.IPv4 {
+			if privateIP(*ip) {
+				addresses = append(addresses, ip.String())
+			}
+		}
+	}
+	return addresses, err
+}
+
+func getLinodeClient(userAgent, apiToken string) linodego.Client {
+	tokenSource := oauth2.StaticTokenSource(&oauth2.Token{AccessToken: apiToken})
+
+	oauth2Client := &http.Client{
+		Transport: &oauth2.Transport{
+			Source: tokenSource,
+		},
+	}
+
+	client := linodego.NewClient(oauth2Client)
+	if userAgent != "" {
+		client.SetUserAgent(userAgent)
```

Comment:
This looks like it is appropriately setting the user agent to me.

Reply:
Before the `!= ""` check was added, I noticed that the `UserAgent` was being cleared (which resulted in a "go-resty" UserAgent (which linodego uses)).

In fact, with 1d3d306 this still does not appear to provide a "go-discover" UserAgent.  The `userAgent` value being supplied is empty.

```sh
$ LINODE_DEBUG=1 ./discover addrs provider=linode region=us-west address_type=public_v4
2019/03/19 09:58:23 [INFO] LINODE_DEBUG being set to true
Registered providers: [aliyun aws azure digitalocean gce linode os packet scaleway softlayer triton vsphere]
[DEBUG] discover: Using provider "linode"
[DEBUG] discover-linode: Using address_type=public_v4 region=us-west tag_name=
RESTY 2019/03/19 09:58:23
---------------------- REQUEST LOG -----------------------
GET  /v4/linode/instances  HTTP/1.1
HOST   : api.linode.com
HEADERS:
                   Accept: application/json
             Content-Type: application/json
               User-Agent: linodego 0.7.0 https://github.com/linode/linodego
                 X-Filter: {"region":"us-west"}
BODY   :
```

Reply:
I followed where `userAgent` was being set and it seemed like it was right on the go-discover side. Maybe you could hard-code it and try to debug why the client is not respecting it?

Reply:
If that comes up empty definitely could be on the go-discover side, but seems right from what I can tell.